src/workerd/jsg/jsg.h

```diff
@@ -649,26 +705,23 @@ namespace {
 // exporting), then you should be able to use '$$' as the identifier prefix in C++ since only the
 // first '$' gets stripped.
 #define JSG_STRUCT(...) \
-  static constexpr ::workerd::jsg::JsgKind JSG_KIND KJ_UNUSED = \
-      ::workerd::jsg::JsgKind::STRUCT; \
-  static constexpr char JSG_FOR_EACH(JSG_STRUCT_FIELD_NAME, , __VA_ARGS__); \
+  JSG_STRUCT_INTERNAL_KIND_AND_FIELDS_AND_TYPE_WRAPPER(__VA_ARGS__); \
   template <typename Registry, typename Self> \
-  static void registerMembers(Registry& registry) { \
+  static void registerMembers(Registry& registry) \
+      requires (!jsg::HasConfiguration<Self>) { \
     JSG_FOR_EACH(JSG_STRUCT_REGISTER_MEMBER, , __VA_ARGS__); \
-    if constexpr (::workerd::jsg::HasStructTypeScriptRoot<Self>::value) { \
-      registry.registerTypeScriptRoot(); \
-    } \
-    if constexpr (::workerd::jsg::HasStructTypeScriptOverride<Self>::value) { \
-      registry.template registerTypeScriptOverride<Self::_JSG_STRUCT_TS_OVERRIDE_DO_NOT_USE_DIRECTLY>(); \
-    } \
-    if constexpr (::workerd::jsg::HasStructTypeScriptDefine<Self>::value) { \
-      registry.template registerTypeScriptDefine<Self::_JSG_STRUCT_TS_DEFINE_DO_NOT_USE_DIRECTLY>(); \
-    } \
+    JSG_STRUCT_REGISTER_MEMBERS_TS_ROOT(); \
+    JSG_STRUCT_REGISTER_MEMBERS_TS_OVERRIDE(); \
+    JSG_STRUCT_REGISTER_MEMBERS_TS_DEFINE(); \
   } \
-  template <typename TypeWrapper, typename Self> \
-  using JsgFieldWrappers = ::workerd::jsg::TypeTuple< \
-    JSG_FOR_EACH(JSG_STRUCT_FIELD, , __VA_ARGS__) \
-  >
+  template <typename Registry, typename Self> \
+  static void registerMembers(Registry& registry, jsg::GetConfiguration<Self> arg) \
+      requires jsg::HasConfiguration<Self> { \
+    JSG_FOR_EACH(JSG_STRUCT_REGISTER_MEMBER, , __VA_ARGS__); \
+    JSG_STRUCT_REGISTER_MEMBERS_TS_ROOT(); \
```

Comment:
Instead of using macros for code sharing here, could the two versions of the method call a common implementation that is a template? These macros are pretty hard to follow -- and they pollute the global scope.

Reply:
yeah I think there is a way to do that, that suggestion makes sense.

Reply:
yeah I think there is a way to do that, that suggestion makes sense.